slack_today_i_did/our_repo.py

```diff
@@ -99,21 +102,82 @@ def get_files_for_017(self, pattern: str) -> List[str]:
             if self.what_kinda_file(filename) == ElmVersion.v_017
         ]
 
-    def get_017_porting_breakdown(self, pattern: str) -> Dict[str, Dict[str, int]]:  # noqa: E501
+    def get_matching_filenames(self, pattern: str) -> List[str]:
         pattern = pattern.replace('.', '/')
         all_files = glob.glob(
             f'{self.repo_dir}/**/{pattern}.elm',
             recursive=True
         )
+        return all_files
+
+    @contextmanager
+    def lock(self) -> None:
+        self.create_lock()
+        yield
+        self.remove_lock()
+
+    def create_lock(self) -> None:
+        """ create a lock so that caches can be used to make some calcs faster """
+        self._breakdown_lock = True
+        self._breakdown_cache = {}
+        self._import_cache = {}
+
+    def remove_lock(self) -> None:
+        """ removes a lock so caches are no longer used """
+        self._breakdown_lock = False
+
+    def get_017_porting_breakdown(self, pattern: str) -> Dict[str, Dict[str, int]]:  # noqa: E501
+        all_files = self.get_matching_filenames(pattern)
 
-        return {
+        breakdown = {
             filename: self.how_hard_to_port(filename) for filename in all_files
             if self.what_kinda_file(filename) == ElmVersion.v_016
         }
 
+        if len(all_files) == 1:
+            imports = self.file_import_list(all_files[0])
+            for import_ in imports:
+                file_names = self.get_matching_filenames(import_)
+
+                if not any(name in breakdown for name in file_names):
+                    import_breakdown = self.get_017_porting_breakdown(import_)
+                    breakdown.update(import_breakdown)
+
+        return breakdown
+
+    def file_import_list(self, filename: str) -> List[str]:
+        """ returns the list of modules imported by a file """
+
+        if self._breakdown_lock:
+            if filename in self._import_cache:
+                return self._import_cache[filename]
+
+        import_lines = []
+
+
+        with open(filename) as f:
+            for line in f:
+                if line.startswith('import'):
+                    without_import = line.lstrip('import ')
+                    just_the_module = without_import.split(' ')[0]
+
+                    import_lines.append(just_the_module)
+                elif not (line.startswith('module') or line.startswith(' ')):
```

Comment:
🐛 There can be a doc comment before the imports, and also imports may not be grouped in a single chunk if the file isn't elm-formatted. ..There can be all sorts of combinations of imports and comments, but for practicality,

```python
elif seen_an_import and line_looks_like_a_type_or_a_function:
```

might do